src/m3/types.ts

```diff
@@ -4,42 +4,56 @@
  */
 
 import {MultiRef, SingleRef, unresolved} from "../references.ts"
-import {Id, Node} from "../types.ts"
+import {Id, Node as NodeInterface} from "../types.ts"
 import {allFeaturesOf} from "./functions.ts"
 
 
 /**
  * Joins fragments of a qualified name using the `.` character.
  */
-export const qualify = (...names: string[]): string =>
-    names.join(".")
+export const qualify = (...names: (string|undefined)[]): string =>
+    names
+        .filter((name) => typeof name === "string")
+        .join(".")
 
 
-interface NamespaceProvider {
+
+abstract class NodeClass implements NodeInterface {
+    parent?: NamespaceProvider  // every parent provides a namespace
```

Comment:
(Assuming the context is every model, not only metamodels.)
I don't think every parent is a NamespaceContainer. I might skip a level when creating namespaces, or create two namespaces for one element in the ancestor axis.

Reply:
That statement is only correct for M2s (as M3-instances). So, “every parent in an M2 also _happens_ to be a namespace” is more accurate. (Only the following can have children/be parents: `Metamodel`, `Concept`, `ConceptInterface`, `Enumeration`.)

Reply:
So `NodeClass` is only used to represent M2 contents?

Reply:
Correct. It could therefore receive a better name. Suggestions? It can't be `M3Concept` because that's an already-defined sum type (which is more precise than an unsealed abstract class).

Reply:
I'd like to push 1 more commit (with some small improvements) after finding a good name for `NodeClass`.

Reply:
Maybe `"M3Node"`? (Or `"M2Node"`?)